Feature/rtl audit login

malinajirka opened the pull request:
Partially fixes #2278 - Fixes RTL layouts on sign in/up screens.

Some Views are not properly positioned in the RTL language mode. 

To test: 
1. Check layouts of the 
WordPress.com login (email) + email password screens
Selft-hosted site login (site URL) + username and password screens
Create Wordpress.com account screen
2. Change your locale to a RTL language (eg. Hebrew) or force RTL layouts in the developer options.
3. Check the layouts, see step 1

Doesn't fix the ViewPager swipe direction in the RTL mode on the first app screen.

khaykov commented:
Great start!

I think the code looks good, but I have a question - there are some changes to state management in `WPLoginInputRow.java`. What are they for?

Also, I think I found a couple of visual issues.

Should "Enter your WordPress.com password" and email label be aligned right?
<img width="415" alt="issue_1" src="https://user-images.githubusercontent.com/728822/35479754-027fc380-0442-11e8-936f-de31f8760082.png">

Error message shouild "point" at the email on the right.
<img width="483" alt="issue_2" src="https://user-images.githubusercontent.com/728822/35479759-24dde6be-0442-11e8-9454-d6715555b32e.png">

This one applies to the majority of input fields - text should start from the right.
<img width="413" alt="issue_3" src="https://user-images.githubusercontent.com/728822/35479765-5f05d22a-0442-11e8-914a-52471a192c70.png">

<img width="416" alt="issue_5" src="https://user-images.githubusercontent.com/728822/35479767-6a18f4ee-0442-11e8-9662-31369202aad7.png">

Thanks!

khaykov commented on libs/login/WordPressLoginFlow/src/main/java/org/wordpress/android/login/widgets/WPLoginInputRow.java:
```diff
@@ -3,12 +3,15 @@
+import android.os.Bundle;
 import android.os.Parcel;
 import android.os.Parcelable;
+import android.support.annotation.NonNull;
```
I think we can safely remove this import.

malinajirka replied:
Fixed;).

malinajirka commented:
Thanks for the feedback @khaykov !

> there are some changes to state management in WPLoginInputRow.java. What are they for?

I've removed the method 'reassignIds()' which was changing View ids so multiple instances of the compound View can exist in the same layout. I wasn't able to make it work with RTL layouts. 

The RelativeLayout class has two fields mRules and mInitialRules. They are used to store relations between the views. The reassignIds() method was fixing the ids in the mRules array. However, the issue was that the mInitialRules field is private (without a getter) and the reassignIds method left old ids in that array. 

> Should "Enter your WordPress.com password" and email label be aligned right?
I've fixed the email label just in the other branch - my bad, thanks. 

However regarding the "Enter your WordPress.com password" label - I'm still not sure what approach should we take. It's a static text (not entered by a user) and the text is in a LTR language. Should we align all texts in the application manually using the "textAlignment" attribute? If so, what about to set the textAlignment in the theme and fix just the text which aren't supposed to be aligned to start? WDYT?

>This one applies to the majority of input fields - text should start from the right.
I'm a bit confused as it seems to work properly on my emulator. On what api level did you test it on?

>Error message shouild "point" at the email on the right.

Great catch! ;) I'll take a look at that.

Thanks;)

malinajirka commented:
>>Error message should "point" at the email on the right.

>Great catch! ;) I'll take a look at that.

In the LTR mode the arrow points to the "i" icon on the right side. In the RTL mode the arrow is correctly on the left side, but the popup windows is not correctly positioned. It seems like a bug in the Support Library. I've tried to create an empty project, and the behaviour is the same. 
Creating a custom EditText seems like an overkill for such a small issue. WDYT?

khaykov commented:
> I've removed the method 'reassignIds()' which was changing View ids so multiple instances of the compound View can exist in the same layout. I wasn't able to make it work with RTL layouts.
> 
> The RelativeLayout class has two fields mRules and mInitialRules. They are used to store relations between the views. The reassignIds() method was fixing the ids in the mRules array. However, the issue was that the mInitialRules field is private (without a getter) and the reassignIds method left old ids in that array.

Got it, thanks!

> However regarding the "Enter your WordPress.com password" label - I'm still not sure what approach should we take. It's a static text (not entered by a user) and the text is in a LTR language. Should we align all texts in the application manually using the "textAlignment" attribute? If so, what about to set the textAlignment in the theme and fix just texts which aren't supposed to be aligned to start? WDYT?

This one is tricky. I also asked this question myself when working on Reader. You are probably right leaving it as is. Most of the time if the string is localized, it will look correctly if your device locale is set to RTL language. And this is the correct way of addressing this - through localization. I don't think we should override default "textAlignment" and we probably can't patch it here and there, because RTL languages have a different level of localization coverage. I'll try to confirm how it works in calypso and what would be the correct approach for us.

> I'm a bit confused as it seems to work properly on my emulator. On what api level did you test it on?

Weird, it works properly for me now :) Maybe I messed something up. Sorry!

> In the LTR mode the arrow points to the "i" icon on the right side. In the RTL mode the arrow is correctly on the left side, but the popup window is not correctly positioned. It seems like a bug in the Support Library. I've tried to create an empty project, and the behaviour is the same.
> Creating a custom EditText seems like an overkill for such a small issue. WDYT @khaykov ?

Yes, custom EditText is probably overkill. Let's leave this one for later.

Anyway, thanks for the PR! We still have some open questions, but I'm going to merge this one, so we can proceed (and fix lint errors in the main branch).

malinajirka commented:
Perfect, thanks!!
